include/util/alias.hpp

```diff
@@ -0,0 +1,112 @@
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef ALIAS_HPP
+#define ALIAS_HPP
+
+#include <functional>
+#include <iostream>
+#include <type_traits>
+
+namespace osrm
+{
+
+template <typename From, typename Tag> struct Alias;
+template <typename From, typename Tag>
+inline std::ostream &operator<<(std::ostream &stream, const Alias<From, Tag> &inst);
+
+template <typename From, typename Tag> struct Alias final
+{
+    using value_type = From;
+    static_assert(std::is_arithmetic<From>(), "");
+
+    From __value;
+    friend std::ostream &operator<<<From, Tag>(std::ostream &stream, const Alias &inst);
+
+    explicit operator From &() { return __value; }
+    explicit operator From() const { return __value; }
+    Alias operator+(const Alias rhs_) const
+    {
+        return Alias{__value + static_cast<const From>(rhs_)};
+    }
+    Alias operator-(const Alias rhs_) const
+    {
+        return Alias{__value - static_cast<const From>(rhs_)};
+    }
+    Alias operator*(const Alias rhs_) const
+    {
+        return Alias{__value * static_cast<const From>(rhs_)};
+    }
+    Alias operator/(const Alias rhs_) const
+    {
+        return Alias{__value / static_cast<const From>(rhs_)};
+    }
+    Alias operator|(const Alias rhs_) const
+    {
+        return Alias{__value | static_cast<const From>(rhs_)};
+    }
+    Alias operator&(const Alias rhs_) const
+    {
+        return Alias{__value & static_cast<const From>(rhs_)};
+    }
+    bool operator<(const Alias z_) const { return __value < static_cast<const From>(z_); }
+    bool operator>(const Alias z_) const { return __value > static_cast<const From>(z_); }
+    bool operator<=(const Alias z_) const { return __value <= static_cast<const From>(z_); }
+    bool operator>=(const Alias z_) const { return __value >= static_cast<const From>(z_); }
+    bool operator==(const Alias z_) const { return __value == static_cast<const From>(z_); }
+    bool operator!=(const Alias z_) const { return __value != static_cast<const From>(z_); }
+};
+
+template <typename From, typename Tag>
+inline std::ostream &operator<<(std::ostream &stream, const Alias<From, Tag> &inst)
+{
+    return stream << inst.__value;
+}
+
+template <typename T>
+struct is_valid_alias
+    : std::integral_constant<bool,
+                             std::is_trivial<T>::value && std::is_standard_layout<T>::value &&
+                                 std::is_pod<T>::value>
+{
+};
```

Comment:
Hmm why not e.g. (not tested)

    template <typename T, typename Alias>
    struct is_alias_for : is_same<typename Alias::From, T> { };

    template <typename T, typename Alias>
    void assert_alias_for( static_assert(is_alias_for<T, Alias>, ""); );

Then we can check e.g.

    assert_alias_for<int32_t, OsmId>();